app/models/stacks_media_token.rb

```diff
@@ -0,0 +1,91 @@
+# to stream a resource.
+#
+# the order of operations might be something like the following example:
+# * a user makes a request to a stacks URL for a streaming resource.
+# * stacks creates a StacksMediaToken and serializes it to its encrypted string form.
+# * the encrypted string form of the token is included as part of a redirect to wowza.
+# * wowza receives the request for the resource stream, with the encrypted string.
+# * wowza calls back to stacks, and passes along: encrypted token string, IP from which
+#   it received the request, id, and file name.
+# * stacks uses StacksMediaToken#verify_encrypted_token_string to determine whether
+#   the encrypted string represents a still-valid token for the given resource.
+# * if it does, it returns 200 OK to wowza, if it doesn't then it returns 403 Forbidden.
+# * wowza serves the stream or not based on the response it gets from stacks.
+# * this checking in wowza can be implemented via a wowza request processing plugin.
+#
+# note that there's no check for whether a token's been used (there's no record of the token
+# minting in memory or DB, and so nothing to mark).
+class StacksMediaToken
+  MAX_TOKEN_AGE = 1.second
+
+  attr_reader :id, :file_name, :user_ip_addr, :timestamp
+
+  def initialize(id, file_name, user_ip_addr)
+    @id = id
+    @file_name = file_name
+    @user_ip_addr = user_ip_addr
+    @timestamp = Time.zone.now
+  end
+
+  def to_hash
+    {
+      id: id,
+      file_name: file_name,
+      user_ip_addr: user_ip_addr,
+      timestamp: timestamp
+    }
+  end
+
+  # though the object is instantiated by the constructor, the string returned by this
+  # method is what "mints" the token for the purposes of checks from other services
+  def to_encrypted_string
+    self.class.encryptor.encrypt_and_sign to_hash
+  end
+
+  def self.create_from_hash(smt_hash)
+    new smt_hash[:id], smt_hash[:file_name], smt_hash[:user_ip_addr]
+  end
+
+  # this method can throw ActiveSupport::MessageVerifier::InvalidSignature or
+  # ActiveSupport::MessageEncryptor::InvalidMessage (the former can come from calling
+  # MessageVerifier#verify, the latter from MessageEncryptor#_decrypt, both called by
+  # MessageEncryptor#decrypt_and_verify).
+  def self.create_from_encrypted_string(encrypted_string)
+    smt_hash = encryptor.decrypt_and_verify encrypted_string
+    create_from_hash smt_hash
+  end
+
+  def self.encryptor
+    ActiveSupport::MessageEncryptor.new(Rails.application.secrets.secret_key_base)
+  end
+
+  # this is how a string is checked to confirm whether it represents a still-valid token minted by stacks
+  def self.verify_encrypted_token_string(encrypted_string, id, file_name, user_ip_addr)
+    unverified_smt = StacksMediaToken.create_from_encrypted_string encrypted_string
+
+    (
+      unverified_smt.id == id &&
+      unverified_smt.file_name == file_name &&
+      unverified_smt.user_ip_addr == user_ip_addr &&
+      (unverified_smt.timestamp >= MAX_TOKEN_AGE.ago)
```

Comment:
to my eyes, this is the only check for the timestamp validity -- is it newer than 1 second ago.  Is that a reasonable time frame?